frontend/components/sections/BitTableSection.tsx

```diff
@@ -111,55 +100,140 @@ function Bit({ type, driverSizeColor, sizes }: GroupedBit) {
+                        flexWrap="wrap"
+                        gap={1}
+                        justifyContent="flex-end"
+                     >
+                        {sizesForDriver.map((size) => (
+                           <Flex
+                              key={size}
+                              px="1.5"
+                              h="4"
+                              gap="1"
+                              bg="gray.100"
+                              borderRadius="full"
+                              borderColor="gray.200"
+                              borderWidth={1}
+                              alignItems="center"
+                              justify="center"
+                              minW="9"
+                           >
+                              <Text
+                                 as="span"
+                                 color="gray.800"
+                                 fontSize="xs"
+                                 fontFamily="mono"
+                              >
+                                 {size}
+                              </Text>
+                              <Box
+                                 w="1.5"
+                                 h="1.5"
+                                 bg={`${driverSizeColors[driverSize]}.500`}
+                                 borderRadius="full"
+                              ></Box>
+                           </Flex>
+                        ))}
+                     </Flex>
+                  );
+               })}
+            </Flex>
+         </Flex>
       </Flex>
    );
 }
+
+const groupScrewdriverBits = (
+   bits: ScrewdriverBit[]
+): {
+   groupedBits: GroupedBit[];
+   driverSizeColors: Record<string, string>;
+} => {
+   let colorIndex = 0;
+   const driverSizeColors: Record<string, string> = {};
+   const predefinedColors = ['amber', 'violet', 'green'];
+   const grouped: Record<string, GroupedBit> = {};
+
+   bits.forEach((bit) => {
+      const { type, size } = bit;
+
+      if (!driverSizeColors[type.driverSize]) {
+         driverSizeColors[type.driverSize] =
+            colorIndex < predefinedColors.length
+               ? predefinedColors[colorIndex++]
+               : 'brand';
+      }
+
+      let groupItem = grouped[type.name];
+
+      if (!groupItem) {
+         groupItem = {
+            type: {
+               id: type.id,
+               icon: type.icon,
+               name: type.name,
+            },
+            sizes: {},
+            totalBitCount: 0,
+         };
+
+         grouped[type.name] = groupItem;
+      }
+
+      if (!groupItem.sizes[type.driverSize]) {
+         groupItem.sizes[type.driverSize] = size ? [size] : [];
+         groupItem.totalBitCount = size ? 1 : 0;
+      } else if (size != null) {
+         groupItem.sizes[type.driverSize].push(size);
+         groupItem.totalBitCount++;
+      }
+   });
+
+   return {
+      groupedBits: Object.values(grouped),
+      driverSizeColors,
+   };
+};
```

Comment:
I took a crack at simplifying this logic; here's my try: https://github.com/iFixit/react-commerce/compare/bit-table-grouping-rewrite

Reply:
I'm not a fan of lodash but your version is good to read so I'm modifying the implementation to match it.
I'll just tweak a little bit the function in order to handle edge cases like those of two driver sizes without bit size correctly:
<img width="494" alt="Screenshot 2023-12-05 at 16 20 25" src="https://github.com/iFixit/react-commerce/assets/7805759/15c2bb08-c001-4dcc-b79a-5845430192e5">